luavm/src/lib/lua_values/mod.rs

```diff
@@ -0,0 +1,754 @@
+mod lua_table;
+mod tagging;
+
+use self::{lua_table::LuaTable, tagging::*};
+use errors::LuaError;
+use gc::{Finalize, Gc, Trace};
+
+/// Represents a value in Lua.
+#[derive(Debug)]
+pub struct LuaVal {
+    encoded_ptr: usize,
+    /// whether this should drop the pointer or not
+    owns: bool,
+}
+
+impl Finalize for LuaVal {}
+unsafe impl Trace for LuaVal {
+    custom_trace!(this, {
+        // only tables have a Gc inside, so there is no need to mark anything else
+        if let LuaValKind::TABLE = this.kind() {
+            mark(unsafe { &*table_ptr(this.encoded_ptr) });
+        }
+    });
+}
+
+impl LuaVal {
+    /// Create an empty LuaVal which is equivalent to Nil.
+    pub fn new() -> LuaVal {
+        LuaVal {
+            encoded_ptr: 0,
+            owns: true,
+        }
+    }
+
+    /// Copies the pointer into a new LuaVal.
+    fn shallow_copy(&self) -> LuaVal {
+        LuaVal {
+            encoded_ptr: self.encoded_ptr,
+            owns: false,
+        }
+    }
+
+    /// Returns the type of the value store in the pointer.
+    fn kind(&self) -> LuaValKind {
+        LuaValKind::from(self.encoded_ptr)
+    }
+
+    /// Returns true if this value can be considered a float, or false otherwise.
+    pub fn is_float(&self) -> bool {
+        match self.kind() {
+            LuaValKind::FLOAT | LuaValKind::STR => true,
```

Comment:
Just to check: does Lua automatically coerce strings to floats?

Reply:
Yes, but only when they are part of an arithmetic expression.

Reply:
OK, so is `is_float` only called during arithmetic expressions? We might need to document that?

Reply:
Yes it is only used in add, sub, etc. Sure, I will add more documentation, and also make it private, as it is used internally only.